File: app/src/main/java/io/nekohasekai/sagernet/fmt/trojan/TrojanBean.java
package io.nekohasekai.sagernet.fmt.trojan;

import androidx.annotation.NonNull;
import com.esotericsoftware.kryo.io.ByteBufferInput;
import com.esotericsoftware.kryo.io.ByteBufferOutput;
import io.nekohasekai.sagernet.fmt.KryoConverters;
import io.nekohasekai.sagernet.fmt.v2ray.StandardV2RayBean;
import moe.matsuri.nb4a.SingBoxOptions;

import org.jetbrains.annotations.NotNull;

public class TrojanBean extends StandardV2RayBean {

    public static final Creator<TrojanBean> CREATOR = new CREATOR<TrojanBean>() {
        @NonNull
        @Override
        public TrojanBean newInstance() {
            return new TrojanBean();
        }

        @Override
        public TrojanBean[] newArray(int size) {
            return new TrojanBean[size];
        }
    };
    public String password;

    @Override
    public void initializeDefaultValues() {
        super.initializeDefaultValues();
        if (security == null || security.isEmpty()) security = "tls";
        if (password == null) password = "";
    }

    @Override
    public void serialize(ByteBufferOutput output) {
        output.writeInt(0);
        super.serialize(output);
        output.writeString(password);
    }

    @Override
    public void deserialize(ByteBufferInput input) {
        int version = input.readInt();
        super.deserialize(input); // StandardV2RayBean
        password = input.readString();
    }

    @NotNull
    @Override
    public TrojanBean clone() {
        return KryoConverters.deserialize(new TrojanBean(), KryoConverters.serialize(this));
    }

    @Override
    public @NotNull String outboundType() throws Throwable {
        return SingBoxOptions.TYPE_TROJAN;
    }
}

File: app/src/main/kotlin/fr/husi/fmt/trojan/TrojanBean.kt
package fr.husi.fmt.trojan

import com.esotericsoftware.kryo.io.ByteBufferInput
import com.esotericsoftware.kryo.io.ByteBufferOutput
import fr.husi.fmt.KryoConverters
import fr.husi.fmt.v2ray.StandardV2RayBean

class TrojanBean : StandardV2RayBean() {

    companion object {
        @JvmField
        val CREATOR = object : CREATOR<TrojanBean>() {
            override fun newInstance(): TrojanBean {
                return TrojanBean()
            }

            override fun newArray(size: Int): Array<TrojanBean?> {
                return arrayOfNulls(size)
            }
        }
    }

    var password: String = ""

    override fun serialize(output: ByteBufferOutput) {
        output.writeInt(0)
        super.serialize(output)
        output.writeString(password)
    }

    override fun deserialize(input: ByteBufferInput) {
        input.readInt()
        super.deserialize(input) // StandardV2RayBean
        password = input.readString()
    }

    override fun clone(): TrojanBean {
        return KryoConverters.deserialize(TrojanBean(), KryoConverters.serialize(this))
    }

}
